File: plugins/claude-code-homeassistant-hermit/src/ha_agent_lab/history.py
"""History API aggregation — pattern detection for ha-analyze-patterns and ha-morning-brief."""
from __future__ import annotations

import fnmatch
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Any

from .artifacts import write_json_artifact
from .ha_api import HomeAssistantClient
from .time_utils import parse_iso

_DEFAULT_DOMAINS = frozenset({"light", "switch", "cover", "climate", "automation"})
_HISTORY_BINARY_SENSOR_CLASSES = frozenset({"motion", "door", "window", "opening", "occupancy"})

# Informational description of the default scope for documentation and tests.
DEFAULT_ENTITY_SCOPE = (
    "light.*", "switch.*", "cover.*", "climate.*", "automation.*",
    "binary_sensor.* (motion/door/window/opening/occupancy)",
)


def select_history_entities(
    normalized: dict[str, Any],
    *,
    override: list[str] | None = None,
) -> list[str]:
    """Return entity IDs for the history fetch.

    When override is given, tokens containing '*' are expanded against the
    entity_index via fnmatch; tokens without '*' are treated as exact IDs.
    Results are de-duped and sorted. Otherwise the default scope applies:
    all entities in light/switch/cover/climate/automation domains, plus
    binary_sensor entities whose device_class is in the event sensor set.
    """
    entity_index: dict[str, Any] = normalized.get("entity_index", {})
    if override is not None:
        out: set[str] = set()
        for token in override:
            if "*" in token:
                out.update(e for e in entity_index if fnmatch.fnmatchcase(e, token))
            else:
                out.add(token)
        return sorted(out)
    result: list[str] = []
    for entity_id, entity in entity_index.items():
        domain = entity_id.split(".", 1)[0]
        if domain in _DEFAULT_DOMAINS:
            result.append(entity_id)
        elif domain == "binary_sensor":
            attrs = entity.get("attributes") or {}
            if attrs.get("device_class") in _HISTORY_BINARY_SENSOR_CLASSES:
                result.append(entity_id)
    return sorted(result)


def aggregate_history(
    history: dict[str, list[dict[str, Any]]],
    requested: list[str],
    *,
    window_start: datetime,
    window_end: datetime,
    include_transitions: bool = False,
) -> dict[str, dict[str, Any]]:
    """Aggregate HA history events per entity.

    Entities in `requested` that are absent from `history` (no events in the
    window) get synthesized zero-count rows so callers always see a complete
    picture. `state_durations` clips event spans to [window_start, window_end].

    When `include_transitions` is True, each aggregate gains a `transitions`
    list: ordered ``{"ts": <iso>, "state": <str>}`` dicts for real state
    changes only (consecutive duplicate states are collapsed).
    """
    ws = window_start.astimezone(UTC)
    we = window_end.astimezone(UTC)
    result: dict[str, dict[str, Any]] = {}

    for entity_id in requested:
        events = history.get(entity_id)
        if events is None:
            row: dict[str, Any] = {
                "event_count": 0,
                "returned": False,
                "hour_histogram": [0] * 24,
                "last_event_iso": None,
                "state_durations": {},
            }
            if include_transitions:
                row["transitions"] = []
            result[entity_id] = row
            continue

        timestamps = [parse_iso(ev.get("last_changed")) for ev in events]

        histogram = [0] * 24
        last_event: datetime | None = None

        for ts in timestamps:
            if ts is not None:
                ts_utc = ts.astimezone(UTC)
                histogram[ts_utc.hour] += 1
                if last_event is None or ts_utc > last_event:
                    last_event = ts_utc

        state_durations: dict[str, float] = {}
        for i, ev in enumerate(events):
            ev_ts = timestamps[i]
            if ev_ts is None:
                continue
            span_start = max(ev_ts.astimezone(UTC), ws)
            next_ts = timestamps[i + 1] if i + 1 < len(timestamps) else None
            span_end = min(next_ts.astimezone(UTC) if next_ts is not None else we, we)
            if span_end > span_start:
                state = ev.get("state", "")
                state_durations[state] = state_durations.get(state, 0.0) + (span_end - span_start).total_seconds()

        agg: dict[str, Any] = {
            "event_count": len(events),
            "returned": True,
            "hour_histogram": histogram,
            "last_event_iso": last_event.isoformat() if last_event else None,
            "state_durations": {s: int(secs) for s, secs in state_durations.items()},
        }

        if include_transitions:
            transitions: list[dict[str, str]] = []
            prev_state: str | None = None
            for ev, ts in zip(events, timestamps):
                state = ev.get("state", "")
                # An event with an unparseable timestamp is skipped without advancing
                # prev_state (a state change hidden behind a bad ts can be lost). HA
                # timestamps are reliably parseable in practice, so we accept this.
                if state != prev_state and ts is not None:
                    transitions.append({"ts": ts.isoformat(), "state": state})
                    prev_state = state
            agg["transitions"] = transitions

        result[entity_id] = agg

    return result


def detect_time_patterns(aggregates: dict[str, dict[str, Any]]) -> list[dict[str, Any]]:
    """Return entities with a dominant single-hour activity peak.

    A pattern requires total events >= 5 and the peak hour accounting for
    > 50% of total events. Synthesized zero-count rows are skipped.
    """
    patterns: list[dict[str, Any]] = []
    for entity_id, agg in aggregates.items():
        if not agg.get("returned"):
            continue
        total = agg.get("event_count", 0)
        if total < 5:
            continue
        histogram = agg.get("hour_histogram", [0] * 24)
        peak_hour = max(range(24), key=lambda h: histogram[h])
        peak_count = histogram[peak_hour]
        if peak_count / total > 0.5:
            patterns.append({
                "entity_id": entity_id,
                "peak_hour": peak_hour,
                "peak_count": peak_count,
                "total": total,
                "description": f"{entity_id} — {peak_count}/{total} events at {peak_hour:02d}:00 UTC",
            })
    return sorted(patterns, key=lambda p: (-p["peak_count"], p["entity_id"]))


def fetch_history_snapshot(
    root: Path,
    client: HomeAssistantClient,
    normalized: dict[str, Any],
    *,
    window_days: int = 7,
    entity_override: list[str] | None = None,
    include_transitions: bool = False,
) -> dict[str, Any]:
    """Fetch, aggregate, and write a history snapshot artifact.

    Writes both a dated file and a fixed-name latest alias under
    `.claude-code-hermit/raw/snapshot-ha-history-{window_days}d-{date}.json`.
    """
    now = datetime.now(UTC)
    window_start = now - timedelta(days=window_days)

    entity_ids = select_history_entities(normalized, override=entity_override)
    history = client.get_history(entity_ids, window_start, now)
    aggregates = aggregate_history(
        history, entity_ids,
        window_start=window_start, window_end=now,
        include_transitions=include_transitions,
    )
    time_patterns = detect_time_patterns(aggregates)

    returned_entities = sorted(eid for eid, agg in aggregates.items() if agg.get("returned"))
    missing_entities = sorted(eid for eid, agg in aggregates.items() if not agg.get("returned"))

    payload: dict[str, Any] = {
        "window_start": window_start.isoformat(),
        "window_end": now.isoformat(),
        "fetched_at": now.isoformat(),
        "requested_entities": entity_ids,
        "returned_entities": returned_entities,
        "missing_entities": missing_entities,
        "event_total": sum(agg["event_count"] for agg in aggregates.values()),
        "entity_aggregates": aggregates,
        "time_patterns": time_patterns,
    }

    write_json_artifact(
        root,
        ".claude-code-hermit/raw",
        f"snapshot-ha-history-{window_days}d",
        payload,
        latest_name=f"snapshot-ha-history-{window_days}d-latest.json",
    )

    return payload

File: plugins/claude-code-homeassistant-hermit/src/history.ts
// WP7 tier 2 port of src/ha_agent_lab/history.py — History API aggregation,
// pattern detection for ha-analyze-patterns and ha-morning-brief.
//
// Representation note: Python emitted timestamps via `datetime.isoformat()`,
// preserving the source offset (e.g. "+02:00") in `transitions[].ts`. JS
// Dates are instants, so all emitted timestamps here are UTC via isoUtc()
// ("+00:00" form). Same instant, normalized offset; no consumer pins the
// offset (artifacts are read by skills/LLM).

import type { HomeAssistantClient } from './ha-api';
import { writeJsonArtifact } from './artifacts';
import { isoUtc, parseIso } from './time-utils';

const DEFAULT_DOMAINS = new Set(['light', 'switch', 'cover', 'climate', 'automation']);
const HISTORY_BINARY_SENSOR_CLASSES = new Set(['motion', 'door', 'window', 'opening', 'occupancy']);

// Informational description of the default scope for documentation and tests.
export const DEFAULT_ENTITY_SCOPE = [
  'light.*',
  'switch.*',
  'cover.*',
  'climate.*',
  'automation.*',
  'binary_sensor.* (motion/door/window/opening/occupancy)',
] as const;

// fnmatch.fnmatchcase subset: *, ?, [seq], [!seq]; everything else literal.
function fnmatchCase(name: string, pattern: string): boolean {
  let re = '';
  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern[i]!;
    if (ch === '*') {
      re += '.*';
    } else if (ch === '?') {
      re += '.';
    } else if (ch === '[') {
      let j = i + 1;
      if (pattern[j] === '!') j++;
      if (pattern[j] === ']') j++;
      while (j < pattern.length && pattern[j] !== ']') j++;
      if (j >= pattern.length) {
        re += '\\[';
      } else {
        let stuff = pattern.slice(i + 1, j).replaceAll('\\', '\\\\');
        if (stuff.startsWith('!')) stuff = `^${stuff.slice(1)}`;
        re += `[${stuff}]`;
        i = j;
      }
    } else {
      re += ch.replace(/[.+^${}()|\\/[\]]/, '\\$&');
    }
  }
  return new RegExp(`^(?:${re})$`, 's').test(name);
}

/**
 * Return entity IDs for the history fetch.
 *
 * When override is given, tokens containing '*' are expanded against the
 * entity_index via fnmatch; tokens without '*' are treated as exact IDs.
 * Results are de-duped and sorted. Otherwise the default scope applies:
 * all entities in light/switch/cover/climate/automation domains, plus
 * binary_sensor entities whose device_class is in the event sensor set.
 */
export function selectHistoryEntities(
  normalized: Record<string, any>,
  options: { override?: string[] | null } = {},
): string[] {
  const entityIndex: Record<string, any> = normalized.entity_index ?? {};
  const override = options.override ?? null;
  if (override !== null) {
    const out = new Set<string>();
    for (const token of override) {
      if (token.includes('*')) {
        for (const entityId of Object.keys(entityIndex)) {
          if (fnmatchCase(entityId, token)) out.add(entityId);
        }
      } else {
        out.add(token);
      }
    }
    return [...out].sort();
  }
  const result: string[] = [];
  for (const [entityId, entity] of Object.entries(entityIndex)) {
    const domain = entityId.split('.', 1)[0]!;
    if (DEFAULT_DOMAINS.has(domain)) {
      result.push(entityId);
    } else if (domain === 'binary_sensor') {
      const attrs = entity.attributes || {};
      if (HISTORY_BINARY_SENSOR_CLASSES.has(attrs.device_class)) result.push(entityId);
    }
  }
  return result.sort();
}

/**
 * Aggregate HA history events per entity.
 *
 * Entities in `requested` that are absent from `history` (no events in the
 * window) get synthesized zero-count rows so callers always see a complete
 * picture. `state_durations` clips event spans to [windowStart, windowEnd].
 *
 * When `includeTransitions` is true, each aggregate gains a `transitions`
 * list: ordered `{"ts": <iso>, "state": <str>}` objects for real state
 * changes only (consecutive duplicate states are collapsed).
 */
export function aggregateHistory(
  history: Record<string, Array<Record<string, any>>>,
  requested: string[],
  options: { windowStart: Date; windowEnd: Date; includeTransitions?: boolean },
): Record<string, Record<string, any>> {
  const { windowStart, windowEnd, includeTransitions = false } = options;
  const ws = windowStart.getTime();
  const we = windowEnd.getTime();
  const result: Record<string, Record<string, any>> = {};

  for (const entityId of requested) {
    const events = history[entityId];
    if (events === undefined) {
      const row: Record<string, any> = {
        event_count: 0,
        returned: false,
        hour_histogram: new Array(24).fill(0),
        last_event_iso: null,
        state_durations: {},
      };
      if (includeTransitions) row.transitions = [];
      result[entityId] = row;
      continue;
    }

    const timestamps = events.map((ev) => parseIso(ev.last_changed));

    const histogram: number[] = new Array(24).fill(0);
    let lastEvent: Date | null = null;

    for (const ts of timestamps) {
      if (ts !== null) {
        histogram[ts.getUTCHours()]! += 1;
        if (lastEvent === null || ts.getTime() > lastEvent.getTime()) lastEvent = ts;
      }
    }

    const stateDurations: Record<string, number> = {};
    for (let i = 0; i < events.length; i++) {
      const evTs = timestamps[i];
      if (evTs === null) continue;
      const spanStart = Math.max(evTs.getTime(), ws);
      const nextTs = i + 1 < timestamps.length ? timestamps[i + 1]! : null;
      const spanEnd = Math.min(nextTs !== null ? nextTs.getTime() : we, we);
      if (spanEnd > spanStart) {
        const state: string = events[i]!.state ?? '';
        stateDurations[state] = (stateDurations[state] ?? 0) + (spanEnd - spanStart) / 1000;
      }
    }

    const agg: Record<string, any> = {
      event_count: events.length,
      returned: true,
      hour_histogram: histogram,
      last_event_iso: lastEvent ? isoUtc(lastEvent) : null,
      // Python int() truncates toward zero (durations are non-negative).
      state_durations: Object.fromEntries(
        Object.entries(stateDurations).map(([state, secs]) => [state, Math.trunc(secs)]),
      ),
    };

    if (includeTransitions) {
      const transitions: Array<{ ts: string; state: string }> = [];
      let prevState: string | null = null;
      for (let i = 0; i < events.length; i++) {
        const state: string = events[i]!.state ?? '';
        const ts = timestamps[i] ?? null;
        // An event with an unparseable timestamp is skipped without advancing
        // prevState (a state change hidden behind a bad ts can be lost). HA
        // timestamps are reliably parseable in practice, so we accept this.
        if (state !== prevState && ts !== null) {
          transitions.push({ ts: isoUtc(ts), state });
          prevState = state;
        }
      }
      agg.transitions = transitions;
    }

    result[entityId] = agg;
  }

  return result;
}

/**
 * Return entities with a dominant single-hour activity peak.
 *
 * A pattern requires total events >= 5 and the peak hour accounting for
 * > 50% of total events. Synthesized zero-count rows are skipped.
 */
export function detectTimePatterns(
  aggregates: Record<string, Record<string, any>>,
): Array<Record<string, any>> {
  const patterns: Array<Record<string, any>> = [];
  for (const [entityId, agg] of Object.entries(aggregates)) {
    if (!agg.returned) continue;
    const total: number = agg.event_count ?? 0;
    if (total < 5) continue;
    const histogram: number[] = agg.hour_histogram ?? new Array(24).fill(0);
    // max(range(24), key=...) — first index with the maximum count.
    let peakHour = 0;
    for (let h = 1; h < 24; h++) {
      if (histogram[h]! > histogram[peakHour]!) peakHour = h;
    }
    const peakCount = histogram[peakHour]!;
    if (peakCount / total > 0.5) {
      patterns.push({
        entity_id: entityId,
        peak_hour: peakHour,
        peak_count: peakCount,
        total,
        description: `${entityId} — ${peakCount}/${total} events at ${String(peakHour).padStart(2, '0')}:00 UTC`,
      });
    }
  }
  return patterns.sort((a, b) =>
    a.peak_count !== b.peak_count
      ? b.peak_count - a.peak_count
      : a.entity_id < b.entity_id
        ? -1
        : a.entity_id > b.entity_id
          ? 1
          : 0,
  );
}

/** The slice of HomeAssistantClient that fetchHistorySnapshot needs (tests inject a stub). */
export interface HistoryClient {
  getHistory(
    entityIds: string[],
    startTime: Date,
    endTime: Date,
  ): Promise<Record<string, Array<Record<string, any>>>>;
}

/**
 * Fetch, aggregate, and write a history snapshot artifact.
 *
 * Writes both a dated file and a fixed-name latest alias under
 * `.claude-code-hermit/raw/snapshot-ha-history-{windowDays}d-{date}.json`.
 */
export async function fetchHistorySnapshot(
  root: string,
  client: HistoryClient | HomeAssistantClient,
  normalized: Record<string, any>,
  options: {
    windowDays?: number;
    entityOverride?: string[] | null;
    includeTransitions?: boolean;
  } = {},
): Promise<Record<string, any>> {
  const { windowDays = 7, entityOverride = null, includeTransitions = false } = options;
  const now = new Date();
  const windowStart = new Date(now.getTime() - windowDays * 86_400_000);

  const entityIds = selectHistoryEntities(normalized, { override: entityOverride });
  const history = await client.getHistory(entityIds, windowStart, now);
  const aggregates = aggregateHistory(history, entityIds, {
    windowStart,
    windowEnd: now,
    includeTransitions,
  });
  const timePatterns = detectTimePatterns(aggregates);

  const returnedEntities = Object.entries(aggregates)
    .filter(([, agg]) => agg.returned)
    .map(([eid]) => eid)
    .sort();
  const missingEntities = Object.entries(aggregates)
    .filter(([, agg]) => !agg.returned)
    .map(([eid]) => eid)
    .sort();

  const payload: Record<string, any> = {
    window_start: isoUtc(windowStart),
    window_end: isoUtc(now),
    fetched_at: isoUtc(now),
    requested_entities: entityIds,
    returned_entities: returnedEntities,
    missing_entities: missingEntities,
    event_total: Object.values(aggregates).reduce((sum, agg) => sum + agg.event_count, 0),
    entity_aggregates: aggregates,
    time_patterns: timePatterns,
  };

  writeJsonArtifact(
    root,
    '.claude-code-hermit/raw',
    `snapshot-ha-history-${windowDays}d`,
    payload,
    `snapshot-ha-history-${windowDays}d-latest.json`,
  );

  return payload;
}
